bluetooth: host: Introduce nRF Connect SDK Bluetooth Host (#27863)
nrfconnect/sdk-nrf: doc/nrf/protocols/bt/host/bluetooth-le-host.rst

Usually, the first step when creating an application is to decide which roles are needed and go from there.

Peripheral role
===============

Most Zephyr-based Bluetooth LE devices will most likely be peripheral-role devices.
This means that they perform connectable advertising and expose one or more GATT services.
After registering services using the :c:func:`bt_gatt_service_register` function, the application typically starts connectable advertising using the :c:func:`bt_le_adv_start` function.

There are several peripheral sample applications available in the tree, such as :ref:`ble_samples` (|NCS|) |sample path|.

Central role
============

Central role may not be as common for Zephyr-based devices as peripheral role, but it is still a plausible one and equally well supported in Zephyr.
Rather than accepting connections from other devices, a central role device scans for available peripheral devices and chooses one to connect to.
Once connected, a central typically acts as a GATT client, first performing discovery of available services and then accessing one or more supported services.

> To initially discover a device to connect to, the application likely uses the :c:func:`bt_le_scan_start` function, waits for an appropriate device to be found (using the scan
> callback), stops scanning using :c:func:`bt_le_scan_stop`, and then connects to the device using :c:func:`bt_conn_le_create`.

Reviewer: ```suggestion
To initially discover a device to connect to, the application likely uses the :c:func:`bt_le_scan_start` function, waits for an appropriate device to be found (using the scan callback), stops scanning using :c:func:`bt_le_scan_stop`, and then connects to the device using :c:func:`bt_conn_le_create`.
```

There are some sample applications for the central role available in the tree, such as :ref:`ble_samples` (|NCS|) |sample path|.

Observer role
=============

An observer role device uses the :c:func:`bt_le_scan_start` function to scan for device, but it does not connect to any of them.
Instead, it uses the advertising data of found devices, combining it optionally with the received signal strength (RSSI).

Broadcaster role
================

> A broadcaster role device uses the :c:func:`bt_le_adv_start` function to advertise specific advertising data, but the type of advertising is non-connectable, that is, another device cannot connect to it.

Reviewer: advertise advertising data?

Connections
===========

Connection handling and the related APIs can be found in the :ref:`Connection Management <bluetooth_connection_mgmt>` section.

Security
========

To achieve a secure relationship between two Bluetooth devices a process called pairing is used.
This process can either be triggered implicitly through the security properties of GATT services, or explicitly using the :c:func:`bt_conn_set_security` function on a connection object.

To achieve a higher security level and protect against Man-In-The-Middle (MITM) attacks, it is recommended to use some out-of-band channel during the pairing.
If the devices have a sufficient user interface this "channel" is the user itself.
The capabilities of the device are registered using the :c:func:`bt_conn_auth_cb_register` function.
The :c:struct:`bt_conn_auth_cb` structure that is passed to this function has a set of optional callbacks that can be used during the pairing.
If the device lacks some feature, the corresponding callback can be set to NULL.

> For example, if the device does not have an input method but does have a display, the ``passkey_entry`` and ``passkey_confirm`` callbacks would be set to NULL, but the
> ``passkey_display`` would be set to a callback capable of displaying a passkey to the user.

Reviewer: ```suggestion
For example, if the device does not have an input method but does have a display, the ``passkey_entry`` and ``passkey_confirm`` callbacks would be set to NULL, but the ``passkey_display`` would be set to a callback capable of displaying a passkey to the user.
```

Depending on the local and remote security requirements and capabilities, there are four possible security levels that can be reached:

* :c:enumerator:`BT_SECURITY_L1` - No encryption and no authentication.
* :c:enumerator:`BT_SECURITY_L2` - Encryption but no authentication (no MITM protection).
* :c:enumerator:`BT_SECURITY_L3` - Encryption and authentication using the legacy pairing method from Bluetooth 4.0 and 4.1.
* :c:enumerator:`BT_SECURITY_L4` - Encryption and authentication using the LE Secure Connections feature available since Bluetooth 4.2.

L2CAP
=====

> L2CAP stands for the Logical Link Control and Adaptation Protocol.
> It is a common layer for all communication over Bluetooth connections.

Reviewer: ```suggestion
The Logical Link Control and Adaptation Protocol (L2CAP) is a common layer for all communication over Bluetooth connections.
```

PDU used in LE Credit Based Flow Control mode and Enhanced Credit Based Flow Control mode.
It contains a SDU segment and additional protocol information.

.. image:: img/l2cap_k_frame_1.drawio.svg
   :width: 45%
   :alt: Representation of a starting K-frame PDU. The PDU is split into three rectangles, the first one being the L2CAP header, its size is four octets and its made of the PDU length and the channel ID.
         The second rectangle represents the L2CAP SDU length, its size is two octets.
         The third rectangle represents the information payload and its size is less or equal to MPS minus two octets.
         The information payload contains the L2CAP SDU.

.. image:: img/l2cap_k_frame.drawio.svg
   :align: right
   :width: 45%
   :alt: Representation of K-frames PDUs after the starting one.
         The PDU is split into two rectangles, the first one being the L2CAP header, its size is four octets and its made of the PDU length and the channel ID.
         The second rectangle represents the information payload and its size is less or equal to MPS.
         The information payload contains the L2CAP SDU.

Relevant Kconfig
----------------

> The following Kconfig options affect the L2CAP payload sizes and whether credit-based channels are
> available.

Reviewer: ```suggestion
The following Kconfig options have an impact on the L2CAP payload sizes.
They also define whether credit-based channels are available.
```
